### PR TITLE
Support `newlines_in_values` CSV option

### LINKED ISSUE #11472: Add support for `newlines_in_values` to `CsvOptions`
### Is your feature request related to a problem or challenge?

I'm trying to read CSVs that include newlines in (quoted) values.

### Describe the solution you'd like

Some googling revealed that this isn't supported currently by the `arrow-csv` crate, whereas that functionality does exist in the C++ ([`ParseOptions::newlines_in_values`](https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N5arrow3csv12ParseOptions18newlines_in_valuesE)) and Python ([`ParseOptions.newlines_in_values`](https://arrow.apache.org/docs/python/generated/pyarrow.csv.ParseOptions.html#pyarrow.csv.ParseOptions.newlines_in_values)) implementations.

Ideally, a `newlines_in_values` field could be added to [`datafusion::common::config::CsvOptions`](https://docs.rs/datafusion/latest/datafusion/common/config/struct.CsvOptions.html) to support this functionality.

Note that the Python docs call out the performance implications of this:

> Setting this to True reduces the performance of multi-threaded CSV reading.

I haven't dug into the implementation, but I imagine it becomes harder to find the right split point for multi-threaded reading (though, it seems not dissimilar to finding the prev/next linebreak, so perhaps not insurmountable...).

### Describe alternatives you've considered

The only alternative I can see would be to preprocess the CSV before feeding it into DF. I haven't explored this option as I imagine it would take a lot of DF plumbing, and it seems valuable to have parity with other arrow CSV packages (C++ and Python, at least).

### Additional context

I was originally planning to report this against the `arrow-rs` repository, but since my use-case is with `datafusion` I decided to report it here. Let me know if this issue would be more appropriate there and I can move/copy it.

### PR DESCRIPTION
**Pull request by connec**
## Which issue does this PR close?

Closes #11472.

## Rationale for this change

This significantly simplifies the UX when dealing with large CSV files that must support newlines in (quoted) values. By default, large CSV files will be repartitioned into multiple parallel range scans. This is great for performance in the common case but when large CSVs contain newlines in values the parallel scan will fail due to splitting on newlines within quotes rather than actual line terminators.

With the current implementation, this behaviour can only be controlled by the session-level `datafusion.optimizer.repartition_file_scans` and `datafusion.optimizer.repartition_file_min_size` settings.

## What changes are included in this PR?

This commit introduces a `newlines_in_values` option to `CsvOptions` and plumbs it through to `CsvExec`, which includes it in the test for whether parallel execution is supported. This provides a convenient and searchable way to disable file scan repartitioning on a per-CSV basis.

I've added `newlines_in_values` using similar conventions to `has_header`, with `CsvOptions` using an `Option<bool>` and a default value coming from `datafusion::common::config::CatalogOptions`.

For now, in the interests of being surgical, I've just added a new argument to [`CsvExec::new`](https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/struct.CsvExec.html#method.new), which is now triggering `clippy::too_many_arguments`. Before going any further I wanted to see if this was overall a good approach, but I'm happy to refactor this into an options struct or similar.

## Are these changes tested?

Yes – a new test has been added alongside the existing tests for file scan repartioning in `datafusion/core/src/datasource/file_format/csv.rs`.

## Are there any user-facing changes?

- *Breaking:* Add public `datafusion::common::config::CatalogOptions::newlines_in_values: bool` field, default: `false`.
- *Breaking:* Add public `datafusion::common::config::CsvOptions::newlines_in_values: Option<bool>` field, default: `None`.
- *Breaking:* Add public `datafusion::datasource::file_format::options::CsvReadOptions::newlines_in_values: bool` field, default: `false`.
- *Breaking:* Add `newlines_in_values: bool` argument to `datafusion::datasource::physical_plan::CsvExec::new`.
- Add public `datafusion::common::config::CsvOptions::with_newlines_in_values` method.
- Add public `datafusion::datasource::file_format::csv::CsvFormat::with_newlines_in_values` method.
- Add public `datafusion::datasource::file_format::options::CsvReadOptions::newlines_in_values` method.
- Add public `datafusion::datasource::physical_plan::CsvExec::newlines_in_values` method.
- Add `newlines_in_values` to relevant proto files.

### CODE REVIEW

File: datafusion/common/src/config.rs
Comment (alamb) [inline datafusion/common/src/config.rs:1673]:
```suggestion
    /// Set true to support newlines in (quoted) values.
```

File: datafusion/core/src/datasource/file_format/csv.rs
Comment (alamb) [inline datafusion/core/src/datasource/file_format/csv.rs:236]:
```suggestion
    /// Set true to support newlines in (quoted) values.
```

**Review by alamb — COMMENTED** [review]
Thank you @connec  -- this PR looks very nice to me (and a very nice first contribution🏅)

I think the only thing missing is an end to end test showing that with this change DataFusion can actually read CSV files with newlines in them

Here are the instructions for such end to end tets https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

Maybe you could add an example data file in https://github.com/apache/datafusion/tree/main/datafusion/core/tests/data 

And the write a test that defined the external table with the appropriate options and read from it

Ideally you should be able to extend one of the existing test files, perhaps 
https://github.com/apache/datafusion/blob/382bf4f3c7a730828684b9e4ce01369b89717e19/datafusion/sqllogictest/test_files/csv_files.slt#L18-L27



cc @2010YOUY01

File: datafusion/common/src/config.rs
Comment (2010YOUY01) [inline datafusion/common/src/config.rs:189]: We need more descriptive comment help for users without context to understand this config option, which should include 1. It's CSV specific 2. Can be overridden by the same config field in `CsvOptions` 3. Its behavior as stated in `CsvOptions`'s comment.

And update in https://github.com/apache/datafusion/blob/main/docs/source/user-guide/configs.md

**Review by 2010YOUY01 — COMMENTED** [review]
Thank you for the new feature support! It is a great UX improvement given 'newlines in quotes' is defined in [CSV standard](https://www.ietf.org/rfc/rfc4180.txt)

I also find this PR's design is good and very easy to read, I think it's good to go with extra integration tests for CSVs with `newlines_in_values` fields as @alamb mentioned.

**Comment by 2010YOUY01** [comment]
I'm trying to provide more context for the rationale for this PR's implementation:

Currently, DataFusion's parallel CSV scan will first split the file range evenly, and adjust the range border by probing `\n` forward.
It's super hard to support `newliens_in_values` in parallel scan in this way: executor have to probe backwards to determine if the current field is qutoed, and it don't know when to stop.

So this PR's approach (explicitly adding one `CsvOption`) is reasonable, maybe in the future we can parse the whole file during planning to find out the correct split points, if it can help performance in some use cases.

**Comment by connec** [comment]
Thanks both for the rapid feedback. I'll take a look and update the PR.

File: datafusion/common/src/config.rs
Comment (connec) [reply datafusion/common/src/config.rs:1673]: The phrasing was a bit awkward initially as I was thinking of the flag as "ensuring support", rather than directly "supporting", since newlines in values are already supported if the file is below the `datafusion.optimizer.repartition_file_min_size`.

I'm not sure if it's worth conveying that detail through these docs, or else to document this as providing support and treat the fact that newlines in values will "just work" for smaller files as an implementation detail that might change.

Comment (connec) [reply datafusion/common/src/config.rs]: I've expanded and normalised the documentation. I've gone for documenting the flag as enabling support for newlines in values, with the fact that newlines in values might work without it left as an implementation-defined detail.

File: datafusion/common/src/config.rs
Comment (connec) [reply datafusion/common/src/config.rs]: I've expanded the documentation. I forgot about the user guide so I'll do that now as well.

**Comment by connec** [comment]
I've made some documentation updates and added/fixed sqllogictests.

There's the new clippy issue around having too many arguments for `CsvExec::new`, which I can either fix with a new `Options` struct of some kind?

**Comment by alamb** [comment]
> There's the new clippy issue around having too many arguments for `CsvExec::new`, which I can either fix with a new `Options` struct of some kind, or suppress and fix in a follow-up.

I think we should suppress the clippy error for now in this PR and we can file a ticket to improve things as a follow on PR 

I would personally suggest a builder pattern like

```rust
let csv_exec = CsvExec::builder()
  .with_has_header(true)
  .with_newlines_in_values(true)
  ...
  .build()?
```

File: datafusion/core/tests/data/newlines_in_values.csv
Comment (alamb) [inline datafusion/core/tests/data/newlines_in_values.csv:2]: 👍

**Review by alamb — COMMENTED** [review]
Thanks @connec  -- this looks good to me 👍 

I think once we have the CI passing this PR is good to go from my perspective.

**Comment by connec** [comment]
I think that should be the fixes needed for CI.

**Comment by alamb** [comment]
> I think we should suppress the clippy error for now in this PR and we can file a ticket to improve things as a follow on PR

I filed https://github.com/apache/datafusion/issues/11565 to track improving the clippy lint

**Review by alamb — APPROVED** [review]
Thanks again @connec  and @2010YOUY01 👌

**Comment by alamb** [comment]
🤔  the slt test seems to be failing on windows (likely due to newline nonsense). 

https://github.com/apache/datafusion/actions/runs/10014892577/job/27695257736

We should probably figure out a way to not run that test on windows. Gah!

Is this something you have time to look at @connec ? If not, I think we could comment out the slt test to get this PR in and then sort out the testing as a follow on PR.

Note to reduce the feedback loop cycle time,  I merged this branch to main and pushed the commit (which I think means the CI should automatically run now on new commits). Once we have merged this PR I think your subsequent PRs will also have the CI run automatically.

**Comment by connec** [comment]
I'd guess it might be related to there being a `\r` on windows? I have some time to look so I'll see if I can resolve it.

**Comment by connec** [comment]
Unfortunately it seems like the workflows still need approval @alamb. I'm not entirely sure if `.gitattributes` change will fix the slts (I don't have a Windows system to test on) but 🤞

**Comment by connec** [comment]
The issue is caused by the line endings in the `newlines_in_values.csv` file being converted to `CRLF` on windows. While either line ending is accepted as a CSV record terminator by default, newlines in values are not normalised at all. This means that the value returned from the query on Windows includes `CRLF` rather than `LF`. Since newlines in the `.slt` file itself *are* normalised, this then leads to the diff.

For now, I've added a `.gitattributes` entry to checkout the `newlines_in_values.csv` file with `LF` line endings always, which resolves the problem on my windows machine at least.

**Comment by alamb** [comment]
🚀 
Thanks again @connec

**Pull request merged**
